### PR TITLE
Added ability to allocate beeper as a resource.

### PR DESCRIPTION
**Pull request by blckmn**
This is exciting (but dangerous). The ability to move resources around :)

### CODE REVIEW

**Comment by blckmn** [comment]
@ledvinap @hydra not sure if this is what you envisaged... but it is working. Obviously what you can do in CLI you can conceivably do using MSP (I know the plan is to ditch CLI all together).

File: src/main/io/serial_cli.c [hunk 1]
```diff
@@ -2295,6 +2307,31 @@ static void cliSet(char *cmdline)
...
+                                    changeValue = true;
+                                } else {
+                                    uint8_t port = (*eqptr)-'A';
+                                    if (port < 8) {
+                                        eqptr++;
+                                        pin = atoi(eqptr);
+                                        if (pin < 16) {
+                                            tmp.int_value = DEFIO_TAG_MAKE(port, pin);
+                                            ioRec_t *rec = IO_Rec(IOGetByTag(tmp.int_value)); 
+                                            if (rec && rec->owner == OWNER_FREE) {
```
Comment (ledvinap) [inline src/main/io/serial_cli.c:2325]: This will cause problems when user wants to swap pins - reset will be necessary ..

**Comment by ledvinap** [comment]
Well .. in the end I would prefer to define pin function (for all user-accessible pins), not pin used for some peripheral. But in meantime this will work fine.

**Comment by blckmn** [comment]
So you'd prefer a cli command such as: pin B7=Beeper

**Comment by ledvinap** [comment]
That won't work well till all drivers support it. And cli is quite cumbersome for it.

The idea is to have list of all available pins in configurator and assign function for each one.

Configurator can check if configuration is valid (checking resource conflicts) and firmware can only check conflicts during initialization, aborting and restoring default configuration on first problem.
Current conflict resolution code is quite complicated and there is no way to inform user what caused the problem.

**Comment by blckmn** [comment]
Hmmm I'm not so sure on that one. I would prefer to configure a peripheral in the one spot, including allocate pins to it. Why? Because most likely you will be configuring other options for that peripheral too not just pins.

Usability is already a pain flicking between ports and config just to configure serial rx. Which one you set first is not clear until you experiment. We need to think about what a user tries to achieve. Setting serial rx, I should specify my com port from a list of uarts already setup or a pop up allowing me to configure new uart if one is available but not yet setup (pin setup would apply here), set the protocol, set the inversion, and save.

So to me I think of it from the perspective of configuring a peripheral not a pin. The inverse to your preference I guess.

**Comment by ledvinap** [comment]
Configurator can present view you describe to user.

It has resources to check complete configuration and resolve conflicts (informing user / allowing user to select appropriate action).
In the end it can generate table with pin mapping and upload it info firmware. (And possibly visualize it for user)

**Comment by rs2k** [comment]
I suggest we look into removing CLI after the first stable release of raceflight... which should be soon. I'm still working out a few flight issues.

I like OP's method of mapping the hardware. It's basic, but is pretty intuitive. An image of the board with drop downs and some basic logic in the configurator won't be too difficult to create.

**Comment by blckmn** [comment]
Ok. I'll give it some more thought. The way the drivers are currently setup, it'll mean a bit of work. Essentially need a pinMap, of ioTag and function. Similar though to ioRec (which has owner - but is not enough granularity to identify function). So both maps are unnecessary duplication. So perhaps it's simply a case of changing IOInit -> IOGetByFunction(function) to get the functions pin rather than set owner (which is actually already set by uploading map through Configurator). Then change owners to functions i.e. I2C1_SCL, I2C1_SDA, BEEPER, UART1_RX etc. if this comes back as NONE / NULL then the function knows it is disabled and can act accordingly.

So we create a means by which to store and reload in masterConfig the ioRecs / function map.

Then in beeperInit I can simply call:

IO_t beeperIO = IOGetByFunction(BEEPER); /\* removing IOInit() */

performing the rest of the initialisation as per normal. If it returns IO_NONE then beeper is disabled.

**Comment by ledvinap** [comment]
@blcm: I implemented owner mainly for debugging purposes - it is easier to see what is pin used for (and debug resource conflicts).Resources were planed for conflict resolution.

Some situations are quite complicated for automatic configuration -
For example softserial may use second timer channel to optimize performance, but the  pin associated with channel (or pins in case od AF map on F3) is free for other use.
Allocating pin between Softserial and PWM is chicken and egg problem - softserial should claim pin only if it is unused, but PWM claims all pins available.
Also there are 4 ADCs on F3 - one pin may me connected to multiple ADCs, which one to choose? Split channels evenly? Or tray to use as few ADCs as possible.
Both is much easier when user is responsible for correct decision.

Mapping `<IO> -> driver<idx>_<pin> as you describe makes sense. Driver can just check that returned pin is usable for given function.

Thinking of it - assigning IO_Tag to driver about equivalent (USART1: TX:PA6 / RX:PA7 ; RX_REMAP:PX5 TX_REMAP:PX5) (PWM: PA1 PA2 PA3)

**Comment by ledvinap** [comment]
I'll have to finish it in my repo and see how it comes out ...
Would be great if someone can help with configurator side - I don't like js much ...

**Comment by blckmn** [comment]
I think we leave it up to the user to configure. Automatic configuration is overkill.

I was thinking PWM1..8 as resources. PWM1 has both a pin, and a timer. Obviously the timer is dependent on the pin selected - but you don't necessarily need complex conflict resolution in the target - that can be in the configurator. The target can have simple checks - and fail gracefully by disabling that function.

I think a pin only has one function (at a time). USART1_RX, USART1_TX, USART2_RX, PWM1, PPM, INVERTER, BEEPER, I2C1_SCL, I2C1_SDA, SPI1_MISO, SPI1_MOSI, SPI1_SCK, SPI_CSS1, SPI_CS2, SPI_CS3 

The chip selects (the function) for SPI can then be assigned to a driver. e.g. MPU could be assigned SPI_CS3 - and then IO_t mpuCsIO = IOGetByFunction(SPI_CS3);

PWM1 would have 
IO_t pwmIO = IOGetByFunction(PWM1); 
TIMER_t pwmTimer = TIMERGetByFunction(PWM1);

**Comment by blckmn** [comment]
Actually to go one step further it would be MOTOR1, MOTOR2, MOTOR3... and PWM_IN1, PWM_IN2 etc SERVO1, SERVO2, SERVO3 etc

To separate the input and the output.

**Comment by hydra** [comment]
@rs2k i was having similar thoughts regarding a board image and dropdowns to select functionality.  Yes, @blckmn different PWM usages need to be treated separately.  Also some indication in the UI of how tolerant a pin is on the board is likely needed (but can come later).  e.g. 5 volt tolerant, has ESD protection, etc.

**Comment by blckmn** [comment]
Closing this one as I have some more work on it

**Pull request closed without merging**